feat(empty screen): new empty message component

ardasnturk opened the pull request:
"No content found" text added by default on a page without content.

The EmptyPage.js file has been created in the Components folder. Updated IndexScreen.js and HtmlScreen.js pages. 

home screen events widget | no events for the current day | home screen about item at the bottom | no content
-- | -- | -- | --
 ![HomeScreen](https://user-images.githubusercontent.com/1942953/150178917-29fbf1f5-2b35-4504-9eb3-0d772d9a21b7.png) |  ![Simulator Screen Shot - iPhone 13 - 2022-01-21 at 13 33 27](https://user-images.githubusercontent.com/11755668/150515452-7a3a411d-0d74-446b-89e9-7791b17bcbf1.png) | ![HomeScreen](https://user-images.githubusercontent.com/1942953/150178965-cf0f9af1-a58d-4f2c-adb4-a54762fb4e67.png)  |  ![Simulator Screen Shot - iPhone 13 - 2022-01-21 at 13 33 37](https://user-images.githubusercontent.com/11755668/150515184-e252e05a-23db-4145-baa2-3abcdd110b56.png) 
HomeScreen | IndexScreen | HomeScreen | HtmlScreen

Fixes #49 

SVA-493

donni106 commented:
Thanks @ardasnturk for this pull request! 🚀 

At first sight on the screenshots you found the right places, well done. On the other hand, the rendered text is too big and bold and does not fit to our other parts of the app. There is also some top padding missing on the HtmlScreen. We have some components already, that take care of paddings and text sizes correctly according to the whole design of the app. Can you find and use them? I will go through the code soon and point out opportunities at the certain code parts.
Our app supports landscape mode. Please check as well, if the screens look good in shape for that cases.

ardasnturk commented:
It was a little difficult for me to find them, as I had not been given any orientation about the code before. I have done my best so far, but if you explain the code beforehand, it will be more explanatory for me. When there was no data in the html page, a blank page appeared before me. That's why I edited the code accordingly. Once again, I will look at the codes and try to arrange them according to the design components.

donni106 commented on src/components/EmptyPage.js:
```diff
@@ -0,0 +1,49 @@
+    orientation === 'landscape' || dimensions.width > consts.DIMENSIONS.FULL_SCREEN_MAX_WIDTH;
+
+  return (
+    <Text
+      h4
+      h4Style={[styles.titleStyle, needLandscapeStyle && styles.titleStyleLandscape]}
+      containerStyle={[styles.containerStyle, needLandscapeStyle && styles.containerStyleLandscape]}
+    >
+      {title}
+    </Text>
```
there are some `Text` components which are the base for every rendered text in the app. good point, that you have found `react-native-elements` already. `RegularText` is something we use often with some different props available. can you try to use some existing text component here?

donni106 commented on src/components/EventList.js:
```diff
@@ -58,6 +60,9 @@ export const EventList = ({
           return <LoadingSpinner loading={!listEndReached} />;
         }
 
+        if (!data?.length) {
+          return <EmptyPage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />;
+        }
```
interesting, works, right? without checking deeper i haven't think of puttin it in the footer. have you tried `ListEmptyComponent` of the react native section list? https://reactnative.dev/docs/sectionlist#listemptycomponent

maybe this fits more in our use case?

donni106 reviewed (commented):
> It was a little difficult for me to find them, as I had not been given any orientation about the code before. I have done my best so far, but if you explain the code beforehand, it will be more explanatory for me. When there was no data in the html page, a blank page appeared before me. That's why I edited the code accordingly. Once again, I will look at the codes and try to arrange them according to the design components.

i understand and know what you mean. you did a good job with finding solutions for the given problem without knowing anything about the big code base.
the pull request is intended now to discuss the best solution we can achieve together.

ardasnturk commented:
I got a build error because of eslint rules. That's why I had to commit a few times. 

Problems solved:

- Renamed EmptyPage.js to `EmptyMessage.js`.
- Instead of the React Native Elements Text Component, the `RegularText` component written inside the application is used.
- `ListEmptyComponent` has been added to the event listing screen.

donni106 commented on src/components/EmptyPage.js:
```diff
@@ -0,0 +1,49 @@
+import { colors, consts, normalize } from '../config';
+import { OrientationContext } from '../OrientationProvider';
+
+export const EmptyPage = ({ title }) => {
```
It will be `EmptyMessage` now, which works perfectly, as it really just contains the message for empty stuff.

donni106 commented:
> I got a build error because of eslint rules. That's why I had to commit a few times.
> 
> Problems solved:
> 
> * Renamed EmptyPage.js to `EmptyMessage.js`.
> * Instead of the React Native Elements Text Component, the `RegularText` component written inside the application is used.
> * `ListEmptyComponent` has been added to the event listing screen.

@ardasnturk thanks! sounds very good. can you please share screenshots again (portrait and landscape)?

donni106 commented on src/components/EventList.js:
```diff
@@ -35,14 +36,19 @@ export const EventList = ({
   };
 
-  const renderItem = useRenderItem(QUERY_TYPES.EVENT_RECORDS, navigation, { noSubtitle });
+  const renderItem = useRenderItem(QUERY_TYPES.EVENT_RECORDS, navigation, {
+    noSubtitle
+  });
```
this was linted by you because of some eslint problems occurred within your development environment?

ardasnturk replied:
I only got ESlint errors over punctuation marks.

donni106 replied:
reverted this style right now together with you in pair review

donni106 commented on src/components/EventList.js:
```diff
@@ -35,14 +36,19 @@ export const EventList = ({
       return previous;
     } else {
-      previous.push({ title: current.params.details.listDate, data: [current] });
+      previous.push({
+        title: current.params.details.listDate,
+        data: [current]
+      });
```
this was linted by you as well because of some eslint problems occurred within your development environment?

ardasnturk replied:
I only got ESlint errors over punctuation marks.

donni106 replied:
maybe some linting issue, where the max line length is setup to 80 at your machine? we use 100.
but no problem, i will check this before merge and correct if necessary. same in EventList.js.

donni106 replied:
we will keep the new one for better readability of the object

donni106 commented on src/screens/HomeScreen.js:
```diff
@@ -145,7 +145,6 @@ export const HomeScreen = ({ navigation, route }) => {
           refreshTimeKey="publicJsonFile-homeCarousel"
         />
         <Widgets widgetConfigs={widgetConfigs} />
-
```
why was this line removed here? it was helpful to separate the several sections.

ardasnturk replied:
I guess it's a Syntax issue. I will be more careful about this from now on.

donni106 commented on src/screens/HtmlScreen.js:
```diff
@@ -22,7 +29,8 @@ export const HtmlScreen = ({ navigation, route }) => {
 
-  if (!query || !queryVariables?.name) return null;
+  if (!query || !queryVariables?.name)
+    return <EmptyMessage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />;
```
we try to avoid hardcoded texts and use references to texts from the file `src/config/texts`.
i could think of something like `texts.empty.content` here.
the text itself could be changes to: "Schade, es wurde kein passender Inhalt gefunden, bitte zu einem späteren Zeitpunkt erneut versuchen."

ardasnturk replied:
I will update the code in this way.

donni106 replied:
you have missed this one, right? it should be a different text here as for the lists.

ardasnturk replied:
Yes true. Sorry, I thought it was the same text. I'm updating right now.

donni106 commented on src/screens/IndexScreen.js:
```diff
@@ -248,6 +250,11 @@ export const IndexScreen = ({ navigation, route }) => {
                 }
+                ListEmptyComponent={
+                  listItems ? (
+                    <EmptyMessage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />
```
as mentioned already, we try to avoid hardcoded texts and use references to texts from the file `src/config/texts`.
i could think of something like `texts.empty.list` here.
the text itself could be changes to: "Schade, es wurden keine passenden Einträge gefunden, bitte zu einem späteren Zeitpunkt erneut versuchen."

ardasnturk replied:
I will update the code in this way.

ardasnturk commented on src/components/EmptyMessage.js:
```diff
@@ -0,0 +1,20 @@
+  title: PropTypes.string.isRequired
+};
+
+EmptyMessage.defaultProps = {
+  title: ''
+};
```
You're right, it should be required.

ardasnturk commented:
Update details: 

- Added `text.empty.list` object to `texts.js`,
- Syntax bug fixed,
- Removed defaultProps in `EmptyMessage.js`,
- Message text has been changed.

EventScreen Portrait | EventScreen Landscape | HtmlScreen Portrait | HtmlScreen Landscape
-- | -- | -- | --
| ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 46 59](https://user-images.githubusercontent.com/11755668/150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 03](https://user-images.githubusercontent.com/11755668/150785640-7ba59a2e-4e3a-4750-b76a-b3bcdfe4e3a9.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 17](https://user-images.githubusercontent.com/11755668/150785656-4cc79169-d08a-4982-a076-36131c3c1813.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 20](https://user-images.githubusercontent.com/11755668/150785683-253bbb69-2e22-4c49-bf3f-f495f53d4003.png)
EventScreen | EventScreen | HtmlScreen | HtmlScreen

ardasnturk commented:
Update Details:

- Added `texts.empty.content` to `texts.js` file,
- Changed `HtmlScreen.js` empty content message.

donni106 commented on src/screens/IndexScreen.js:
```diff
@@ -248,6 +250,7 @@ export const IndexScreen = ({ navigation, route }) => {
                     <DropdownHeader {...{ query, queryVariables, data, updateListData }} />
                   ) : null
                 }
+                ListEmptyComponent={listItems ? <EmptyMessage title={texts.empty.list} /> : null}
```
istn't `ListEmptyComponent` implicitly checking the data for presence and therefore we could get rid of `listItems ?...` and pass our `EmptyMessage` directly?

ardasnturk replied:
At the moment, I tried it as you said. I did not encounter any error. This is how I am sending an update.

ardasnturk replied:
Should I use it with `<SafeAreaViewFlex>` ?

donni106 replied:
as the list is rendered already inside the wrappers, i think we do not need anything here.

donni106 commented on src/screens/HtmlScreen.js:
```diff
@@ -22,7 +29,7 @@ export const HtmlScreen = ({ navigation, route }) => {
   const trackScreenViewAsync = useTrackScreenViewAsync();
 
-  if (!query || !queryVariables?.name) return null;
+  if (!query || !queryVariables?.name) return <EmptyMessage title={texts.empty.content} />;
```
we are missing screen wrapping components here. compared to the the "main" return of this component, there are following wrappers we should add here:

- `<SafeAreaViewFlex>`
- `<WrapperWithOrientation>`
- `<Wrapper>`

this way, the empty message should be placed more beautiful on the screen.
can you give it a try? otherwise i will dig deeper before merging the branch once the next release will be drafted.

ardasnturk replied:
I didn't know I had to write it this way. I'm updating the code right now.

donni106 commented:
> Update Details:
> 
> * Removed `ListItems` control in `IndexScreen.js`,
> * `HtmlScreen.js` blank text wrapped with:
> * * `<SafeAreaViewFlex>`
> * * `<WrapperWithOrientation>`
> * * `<Wrapper>`

sounds good. and how does it look like now? are you satisfied with the new outcome?

ardasnturk commented:
> > Update Details:
> 
> > 
> 
> > * Removed `ListItems` control in `IndexScreen.js`,
> 
> > * `HtmlScreen.js` blank text wrapped with:
> 
> > * * `<SafeAreaViewFlex>`
> 
> > * * `<WrapperWithOrientation>`
> 
> > * * `<Wrapper>`
> 
> 
> 
> sounds good. and how does it look like now? are you satisfied with the new outcome?

It definitely looks better now.  I'm slowly starting to understand the code.  Thanks.

donni106 commented:
> It definitely looks better now. I'm slowly starting to understand the code. Thanks.

happy to hear this. would you mind sharing screenshots after visual changes, so that i can see whats happening without running the code?

donni106 commented:
@ardasnturk as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.

![150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86](https://user-images.githubusercontent.com/1942953/150827917-268eca75-dcbb-4b99-ad7f-d4c82b20e182.png)

list headers could have an additional check on list items and follow the existing logic just in case that there are data.
what do you think? do you know what i mean?

ardasnturk commented:
> as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> 
> ![150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86](https://user-images.githubusercontent.com/1942953/150827917-268eca75-dcbb-4b99-ad7f-d4c82b20e182.png)
> 
> list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?

My idea is to keep the filter here. 
If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.

ardasnturk commented:
> > It definitely looks better now. I'm slowly starting to understand the code. Thanks.
> 
> happy to hear this. would you mind sharing screenshots after visual changes, so that i can see whats happening without running the code?

EventScreen Portrait | EventScreen Landscape | HtmlScreen Portrait | HtmlScreen Landscape
-- | -- | -- | --
| ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 33 07](https://user-images.githubusercontent.com/11755668/150860402-344efbfc-0068-403e-b663-464d0d922a68.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 33 11](https://user-images.githubusercontent.com/11755668/150860435-341ec224-9881-4f2b-9a40-1978c718659e.png)  | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 32 53](https://user-images.githubusercontent.com/11755668/150860335-3db96ca8-f2de-447a-aebb-0873fce51fed.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 32 57](https://user-images.githubusercontent.com/11755668/150860571-62aa3e85-5a4f-478d-a6c3-37bf096b7f60.png)
EventScreen | EventScreen | HtmlScreen | HtmlScreen

donni106 commented:
> > as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> > ![150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86](https://user-images.githubusercontent.com/1942953/150827917-268eca75-dcbb-4b99-ad7f-d4c82b20e182.png)
> > list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?
> 
> My idea is to keep the filter here. If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.

give it a try in the simulator or on your device. the filter is always there, no matter the filtered data. so if a user accesses the screen, where no data is provided the filter will not bring data anyways. so the user can filter and filter and it always returns zero data. that's why my thoughts were to remove the filter entirely, if there is no data at all. you see what i mean?

donni106 commented on src/components/EmptyMessage.js:
```diff
@@ -0,0 +1,16 @@
+
+export const EmptyMessage = ({ title }) => {
+  return (
+    <RegularText small center>
```
could setting the prop `placeholder` as well makes sense here to reduce strength of the text?

ardasnturk replied:
Sure, I can try. Should I leave the size as `small`?

donni106 replied:
good question, ähm, yes, `small` makes quite sense to me.

ardasnturk commented:
> > > as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> > > ![150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86](https://user-images.githubusercontent.com/1942953/150827917-268eca75-dcbb-4b99-ad7f-d4c82b20e182.png)
> > > list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?
> > 
> > 
> > My idea is to keep the filter here. If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.
> 
> give it a try in the simulator or on your device. the filter is always there, no matter the filtered data. so if a user accesses the screen, where no data is provided the filter will not bring data anyways. so the user can filter and filter and it always returns zero data. that's why my thoughts were to remove the filter entirely, if there is no data at all. you see what i mean?


I guess you want such a design. I updated the codes in this way. I don't want to post it to github without your consent. I hope I understood correctly.

Empty Event List | Full Event List | 
-- | -- | 
![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 39](https://user-images.githubusercontent.com/11755668/151261733-0dbd66c6-2af4-4dd5-8e99-38536ad95e37.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 32](https://user-images.githubusercontent.com/11755668/151261964-c25396b7-65ed-42fd-ab2b-5612413b7ab6.png)
IndexScreen | IndexScreen |

donni106 commented:
> I guess you want such a design. I updated the codes in this way. I don't want to post it to github without your consent. I hope I understood correctly.
> 
> | Empty Event List | Full Event List | 
> | -- | -- | 
> | ![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 39](https://user-images.githubusercontent.com/11755668/151261733-0dbd66c6-2af4-4dd5-8e99-38536ad95e37.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 32](https://user-images.githubusercontent.com/11755668/151261964-c25396b7-65ed-42fd-ab2b-5612413b7ab6.png)
> | IndexScreen | IndexScreen |

looks very good. i think this is the better user experience!

donni106 commented on src/components/EmptyMessage.js:
```diff
@@ -1,13 +1,19 @@
-    <RegularText small center>
-      {title}
-    </RegularText>
+    <SafeAreaViewFlex>
+      <WrapperWithOrientation>
+        <Wrapper>
```
why is this needed now and moved here? does it still fit for our cases?

donni106 replied:
all right, now i understand the reason. then i also think it is better be placed here, thanks.

ardasnturk replied:
If wrapping is deleted | If wrapping is not deleted |
|--|--
![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 27 05](https://user-images.githubusercontent.com/11755668/151340776-0eee5aeb-0d98-465d-9a8d-6d95d6a08ada.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 18](https://user-images.githubusercontent.com/11755668/151340999-30d73133-4284-4d6f-b161-5872be0e7e01.png) |
IndexScreen.js | IndexScreen.js |

donni106 commented on src/screens/IndexScreen.js:
```diff
@@ -253,6 +253,7 @@ export const IndexScreen = ({ navigation, route }) => {
                 ListEmptyComponent={<EmptyMessage title={texts.empty.list} />}
                 navigation={navigation}
                 data={listItems}
+                noEvent={listItems.length > 0 ? false : true}
```
i think this new prop can be avoided and does not need to be passed down. at the end there is `ListHeaderComponent={noEvent ? null : ListHeaderComponent}` now. so i have in my mind to directly avoid passing something as `ListHeaderComponent` here if there is no data instead of a new prop to compare later, you know?

ardasnturk replied:
Yes I understand that I don't need the extra prop.

ardasnturk commented:
Update Details:

- Deleted `noEvent` prop,
- `ListHeaderComponent` is associated with data.

Screenshots:

IndexScreen.js Portrait | IndexScree.js Landscape |
|--|--
![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 18](https://user-images.githubusercontent.com/11755668/151340999-30d73133-4284-4d6f-b161-5872be0e7e01.png) | ![Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 23](https://user-images.githubusercontent.com/11755668/151342008-366eed3d-8f0c-4362-8bb8-3870bdc940e9.png) |
IndexScreen.js | IndexScreen.js |

donni106 commented:
we could think of reusing the image from #440 for empty screens.

ardasnturk commented:
Update Details:

- Added image for empty event page
- Changed text and image to placeholder

Screenshots:

IndexScreen.js Portrait | IndexScree.js Landscape |
|--|--
![Simulator Screen Shot - iPhone 13 - 2022-03-03 at 09 15 24](https://user-images.githubusercontent.com/11755668/156524277-9dc53f46-34ce-4bdd-8ed6-b3485320bb4a.png) | ![Simulator Screen Shot - iPhone 13 - 2022-03-03 at 09 15 16](https://user-images.githubusercontent.com/11755668/156524307-e8429a37-d3cf-4e3e-8b76-a0012d23dfd3.png)|
IndexScreen.js | IndexScreen.js |

SVA-493
